File: pe-core/src/main/java/org/saar/core/renderer/deferred/shadow/ShadowsScreenPrototype.java
package org.saar.core.renderer.deferred.shadow;

import org.saar.core.screen.ScreenPrototype;
import org.saar.core.screen.annotations.ScreenImageProperty;
import org.saar.core.screen.image.DepthScreenImage;
import org.saar.core.screen.image.ScreenImage;
import org.saar.lwjgl.opengl.constants.DataType;
import org.saar.lwjgl.opengl.constants.DepthFormatType;
import org.saar.lwjgl.opengl.fbos.attachment.DepthAttachment;
import org.saar.lwjgl.opengl.textures.Texture;
import org.saar.lwjgl.opengl.textures.TextureTarget;

public class ShadowsScreenPrototype implements ScreenPrototype {

    private final Texture depthTexture = Texture.create(TextureTarget.TEXTURE_2D);

    @ScreenImageProperty
    private final ScreenImage depthImage = new DepthScreenImage(DepthAttachment.withTexture(
            this.depthTexture, DepthFormatType.COMPONENT24, DataType.U_BYTE));

    public Texture getDepthTexture() {
        return this.depthTexture;
    }
}

File: pe-core/src/main/java/org/saar/core/renderer/shadow/ShadowsScreenPrototype.kt
package org.saar.core.renderer.shadow

import org.saar.core.screen.ScreenPrototype
import org.saar.lwjgl.opengl.constants.InternalFormat
import org.saar.lwjgl.opengl.fbo.attachment.buffer.TextureAttachmentBuffer
import org.saar.lwjgl.opengl.texture.MutableTexture2D
import org.saar.lwjgl.opengl.texture.parameter.TextureMagFilterParameter
import org.saar.lwjgl.opengl.texture.parameter.TextureMinFilterParameter
import org.saar.lwjgl.opengl.texture.parameter.TextureSWrapParameter
import org.saar.lwjgl.opengl.texture.parameter.TextureTWrapParameter
import org.saar.lwjgl.opengl.texture.values.MagFilterValue
import org.saar.lwjgl.opengl.texture.values.MinFilterValue
import org.saar.lwjgl.opengl.texture.values.WrapValue

class ShadowsScreenPrototype : ScreenPrototype {

    private val depthTexture = MutableTexture2D.create().apply {
        applyParameters(
            TextureMinFilterParameter(MinFilterValue.LINEAR),
            TextureMagFilterParameter(MagFilterValue.LINEAR),
            TextureSWrapParameter(WrapValue.CLAMP_TO_EDGE),
            TextureTWrapParameter(WrapValue.CLAMP_TO_EDGE)
        )
    }

    override val depthBuffer = TextureAttachmentBuffer(this.depthTexture, InternalFormat.DEPTH24)

    val buffers = object : ShadowsBuffers {
        override val depth = depthTexture
    }
}